lib/kernel.js

```diff
@@ -16,34 +16,89 @@ import store from "./store";
 import WatchesStore from "./store/watches";
 import OutputStore from "./store/output";
 import HydrogenKernel from "./plugin-api/hydrogen-kernel";
+import InputView from "./input-view";
+import KernelTransport from "./kernel-transport";
+import ZMQKernel from "./zmq-kernel";
+import type { ResultsCallback } from "./kernel-transport";
+
+interface KernelMiddlewareStack {
+  interrupt: () => void;
+  shutdown: () => void;
+  restart: (onRestarted: ?Function) => void;
+  execute: (
+    code: string,
+    callWatches: boolean,
+    onResults: ResultsCallback
+  ) => void;
+  complete: (code: string, onResults: ResultsCallback) => void;
+  inspect: (
+    code: string,
+    cursorPos: number,
+    onResults: ResultsCallback
+  ) => void;
+}
 
 export default class Kernel {
-  @observable executionState = "loading";
   @observable inspector = { bundle: {} };
   outputStore = new OutputStore();
 
-  kernelSpec: Kernelspec;
-  grammar: atom$Grammar;
-  language: string;
-  displayName: string;
   watchesStore: WatchesStore;
-  watchCallbacks: Array<Function> = [];
   emitter = new Emitter();
   pluginWrapper: HydrogenKernel | null = null;
+  kernel: ZMQKernel | KernelTransport; // XXX(nikita): need explicit ZMQKernel for plugins to typecheck
+  middlewareStack: KernelMiddlewareStack;
 
-  constructor(kernelSpec: Kernelspec, grammar: atom$Grammar) {
-    this.kernelSpec = kernelSpec;
-    this.grammar = grammar;
-
-    this.language = kernelSpec.language.toLowerCase();
-    this.displayName = kernelSpec.display_name;
+  constructor(kernel: KernelTransport) {
+    this.kernel = kernel;
 
     this.watchesStore = new WatchesStore(this);
+
+    this.middlewareStack = {
+      interrupt: () => this.kernel.interrupt(),
+      shutdown: () => this.kernel.shutdown(),
+      restart: (onRestarted: ?Function) => this.kernel.restart(onRestarted),
+      execute: (
+        code: string,
+        callWatches: boolean,
+        onResults: ResultsCallback
+      ) => this.kernel.execute(code, callWatches, onResults),
+      complete: (code: string, onResults: ResultsCallback) =>
+        this.kernel.complete(code, onResults),
+      inspect: (code: string, cursorPos: number, onResults: ResultsCallback) =>
+        this.kernel.inspect(code, cursorPos, onResults)
+    };
+  }
+
+  get kernelSpec(): Kernelspec {
+    return this.kernel.kernelSpec;
+  }
+
+  get grammar(): atom$Grammar {
+    return this.kernel.grammar;
+  }
+
+  get language(): string {
+    return this.kernel.language;
+  }
+
+  get displayName(): string {
+    return this.kernel.displayName;
+  }
+
+  setMiddlewareStack(middlewareStack: KernelMiddlewareStack) {
```

Comment:
I wouldn't use the name `Middleware`, I'd use `Plugin`, not only because is shorter, but also because that's the name of the API.

Reply:
Middleware is a standard term [1] [2]. I do not call in a plugin because one hydrogen plugin can create multiple kernel middlewares.

[1] https://docs.djangoproject.com/en/2.0/topics/http/middleware/
[2] https://expressjs.com/en/guide/using-middleware.html

Reply:
There is no reason why a plugin cannot create multiple plugins.

Anyway, this is just a naming issue (I still think that plugin developers will be a bit confused by the need to distinguish plugins from middlewares).